File: src/main/java/dragonbook/inter/stmt/Do.java
package dragonbook.inter.stmt;

import dragonbook.inter.expr.Expr;
import dragonbook.symbols.Type;

public class Do extends Stmt {
    public Expr expr;
    public Stmt stmt;

    public Do() {
        expr = null;
        stmt = null;
    }

    public void init(Stmt s, Expr x) {
        expr = x;
        stmt = s;
        if (expr.type != Type.BOOL) expr.error("boolean required in do");
    }
}

File: src/main/kotlin/dragonbook/inter/stmt/Do.kt
package dragonbook.inter.stmt

import dragonbook.error.ParseError
import dragonbook.inter.expr.Expr
import dragonbook.symbols.Type

data class Do(val stmt: Stmt, val condition: Expr) : Stmt {

    init {
        validate()
    }

    private fun validate() {
        if (condition.type != Type.BOOL) throw ParseError("boolean required in do", condition)
    }

    override fun toString(): String = "do { $stmt } while ($condition)"
}